feat: Adding the option and feature to enable borders with color, opacity and width control on heatmaps along with white borders on emphasis

Dev10-34 opened the pull request:
<!---
Please write the PR title following the conventions at https://www.conventionalcommits.org/en/v1.0.0/
Example:
fix(dashboard): load charts correctly
-->

### SUMMARY
 In heatmaps, when there can be several continuous same colored regions, making it hard to distinguish between each heatmap element itself. Depending on need of the user, this differentiation capability might be required or crucial to their work. 
Added a color picker color control which keeps the borders transparent by default. Users can set the opacity as needed.
Added a slider control to set the border width as needed.  
Added a feature which pops out the element upon emphasis with white borders and shadows.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
BEFORE
![image](https://github.com/user-attachments/assets/de4bddab-faa0-4fe2-9a30-b5d5888953db)



AFTER
![image](https://github.com/user-attachments/assets/fca7da46-c0b5-43a8-bbe8-9392b84bd550)
![image](https://github.com/user-attachments/assets/dd7f652f-e153-4df8-befb-187ad2b21dd0)
![image](https://github.com/user-attachments/assets/b1706255-b894-40cd-8ee6-16133df31137)


### TESTING INSTRUCTIONS
No test files found for the same.

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [x] Changes UI: Adds an 'Enable Borders' Checkbox to Heatmap Customization, adds border and emphasis feature to heatmaps.
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

Dev10-34 committed:
modified:   superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/controlPanel.tsx
	modified:   superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/transformProps.ts
	modified:   superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/types.ts

Dev10-34 committed:
modified:   superset-frontend/package-lock.json
	modified:   superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/transformProps.ts

Dev10-34 committed:
modified:   superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/transformProps.ts

sadpandajoe commented on superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/controlPanel.tsx:
```diff
@@ -278,6 +278,20 @@ const config: ControlPanelConfig = {
+            name: 'enable_borders',
+            config: {
+              type: 'CheckboxControl',
+              label: t('Enable Borders'),
```
This should be sentence case

Dev10-34 replied:
Hello, 
This is my first time creating a PR. What should I be doing next?
Thanking you,
Regards
Dev10-34

michael-s-molina commented:
Thanks for the PR @Dev10-34. Could you instead of adding a checkbox to enabling black borders, add a control to set the border color which by default is transparent (no border)? This is important as the color border may conflict with the color scheme used by the chart.

Dev10-34 commented:
> Thanks for the PR @Dev10-34. Could you instead of adding a checkbox to enabling black borders, add a control to set the border color which by default is transparent (no border)? This is important as the color border may conflict with the color scheme used by the chart.

Thank you for the response! I don't quite get what you mean by a control. Do you mean something like a slider?

michael-s-molina commented:
> Do you mean something like a slider?

Something like the `Series colors `control of the Waterfall chart.
<img width="297" alt="Screenshot 2025-02-24 at 16 30 38" src="https://github.com/user-attachments/assets/05ccb809-dd4a-400c-916a-f680b07ca198" />

Dev10-34 committed:
modified:   superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/controlPanel.tsx
	modified:   superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/transformProps.ts
	modified:   superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/types.ts

Dev10-34 committed:
modified:   superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/controlPanel.tsx
	modified:   superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/transformProps.ts
	modified:   superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/types.ts

Dev10-34 commented:
> > Do you mean something like a slider?
> 
> Something like the `Series colors `control of the Waterfall chart.
> 
> <img alt="Screenshot 2025-02-24 at 16 30 38" width="297" src="https://private-user-images.githubusercontent.com/70410625/416358705-05ccb809-dd4a-400c-916a-f680b07ca198.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA0Nzc2MjAsIm5iZiI6MTc0MDQ3NzMyMCwicGF0aCI6Ii83MDQxMDYyNS80MTYzNTg3MDUtMDVjY2I4MDktZGQ0YS00MDBjLTkxNmEtZjY4MGIwN2NhMTk4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI1VDA5NTUyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY0MTE2MTc5NmI2ZGI1NDE1M2Y0ZGFmOTEyYjIyODhjOTMyNjlmZTQxZWNiZTViNTE2NmM2Mzg1ZTYyODMxM2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.W9jkvJDJA1mWRx1ivvkmji0DV1yMR58LqsuG74FQRNY">

I couldn't quite get the opacity part down. So for now, I have added 2 things. One checkbox which will let you control whether you want the borders or not and secondly a color picker so that you are able to change the color of your border as needed. The opacity function doesn't seem to work for waterfall chart either.

michael-s-molina commented on superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/controlPanel.tsx:
```diff
@@ -224,6 +224,20 @@ const config: ControlPanelConfig = {
+            name: 'element_border',
+            config: {
+              type: 'ColorPickerControl',
+              label: t('Border Color'),
```
```suggestion
              label: t('Border color'),
```

Dev10-34 replied:
Ok. On it

michael-s-molina commented on superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/controlPanel.tsx:
```diff
@@ -224,6 +224,20 @@ const config: ControlPanelConfig = {
+              label: t('Border Color'),
+              renderTrigger: true,
+              description: t(
+                'Select border color for Heatmap elements. Initial opacity is 0. Remember to change as per need',
```
```suggestion
                'Select border color for Heatmap elements. Initial opacity is 0.',
```

Dev10-34 replied:
I am sorry. I  made another mistake. Need to remove "Initial opacity set to 0" as changing opacity bar (a value) doesn't seem to have any effect

michael-s-molina reviewed (commented):
Could you also remove the `package-lock.json` changes as there was no package change in this PR?

Dev10-34 committed:
modified:   superset-frontend/package-lock.json

Dev10-34 committed:
modified:   superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/controlPanel.tsx

Dev10-34 committed:
modified:   superset-frontend/packages/superset-ui-core/src/color/utils.ts
	modified:   superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/transformProps.ts

Dev10-34 commented:
I added a function in utils.ts which lets you change the opacity as well now

michael-s-molina commented on superset-frontend/packages/superset-ui-core/src/color/utils.ts:
```diff
@@ -120,3 +120,29 @@ export function rgbToHex(red: number, green: number, blue: number) {
   return `#${r}${g}${b}`;
 }
+
+export function rgbToHexWithAlpha(
```
We already have the `addAlpha` function for this purpose.

michael-s-molina commented on superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/controlPanel.tsx:
```diff
@@ -224,6 +225,31 @@ const config: ControlPanelConfig = {
+        ],
+        [
+          {
+            name: 'enable_border',
```
If the border color can be transparent by default, you don't need a checkbox to enable the border, you just need to set a color. You could also have a control for border width with a default value of 0 indicating no border.

Dev10-34 replied:
Thank you for being so patient and helpful! I have added the changes as you mentioned. Earlier I didn't realize how to exactly use the addAlpha() function that was defined. I was creating a separate variable for opacity and passing that through the function which ended up not letting it function it properly. Passing elementBorder.a as the second argument was the way to go. 
I removed the function from ui-core packages along with the enable border checkbox. 
I added a slider control for border width as that looked nice. However, I set its default width to 1 as the effects should be visible upon increasing the opacity immediately and the user wouldn't need to work with 2 controls for their effect to show. 
I have a suggestion as well. Instead of having the opacity set to 0 (transparent), I keep the opacity to max, but keep the border width to 0. My reasoning is that for someone who doesn't really care for how the borders look, such a user will need to tweak just 1 visible control for them to be able to visualize their needs properly. Someone who is looking to customize and style their charts can use all the controls as needed.

Dev10-34 replied:
I have implemented my suggestion for the review. Please let me know what default to keep

michael-s-molina commented on superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/controlPanel.tsx:
```diff
@@ -224,6 +225,31 @@ const config: ControlPanelConfig = {
+        ],
+        [
+          {
+            name: 'enable_border',
```
If the border color can be transparent by default, you don't need a checkbox to enable the border, you just need to set a color. You could also have a control for border width with a default value of 0 indicating no border.

Dev10-34 committed:
modified:   superset-frontend/packages/superset-ui-core/src/color/utils.ts
	modified:   superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/controlPanel.tsx
	modified:   superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/transformProps.ts
	modified:   superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/types.ts

Dev10-34 committed:
modified:   superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/controlPanel.tsx

Dev10-34 committed:
modified:   superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/controlPanel.tsx

Dev10-34 committed:
modified:   superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/controlPanel.tsx
	modified:   superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/transformProps.ts

michael-s-molina commented on superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/controlPanel.tsx:
```diff
@@ -19,6 +19,7 @@
 import { t, validateNonEmpty } from '@superset-ui/core';
 import {
   ControlPanelConfig,
+  ControlSubSectionHeader,
```
```suggestion
```

michael-s-molina commented on superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/controlPanel.tsx:
```diff
@@ -224,6 +225,31 @@ const config: ControlPanelConfig = {
         ['y_axis_format'],
         ['x_axis_time_format'],
         ['currency_format'],
+        [<ControlSubSectionHeader>Border</ControlSubSectionHeader>],
```
```suggestion
```

michael-s-molina commented on superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/controlPanel.tsx:
```diff
@@ -224,6 +225,31 @@ const config: ControlPanelConfig = {
         ['currency_format'],
+        [<ControlSubSectionHeader>Border</ControlSubSectionHeader>],
+        [
+          {
+            name: 'element_border',
+            config: {
+              type: 'ColorPickerControl',
+              label: t('Border color'),
+              renderTrigger: true,
+              description: t('Select border color for Heatmap elements'),
+              default: { r: 255, g: 255, b: 255, a: 1 },
+            },
```
Could you also move the configuration after `linear_color_scheme` as it makes sense to be close to the color scheme?

```suggestion
          {
            name: 'border_color',
            config: {
              type: 'ColorPickerControl',
              label: t('Border color'),
              renderTrigger: true,
              description: t('The color of the elements border'),
              default: { r: 0, g: 0, b: 0, a: 1 },
            },
```

michael-s-molina commented on superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/controlPanel.tsx:
```diff
@@ -224,6 +225,31 @@ const config: ControlPanelConfig = {
+              default: { r: 255, g: 255, b: 255, a: 1 },
+            },
+          },
+          {
+            name: 'borders_width',
+            config: {
+              type: 'SliderControl',
+              label: t('Border width'),
+              renderTrigger: true,
+              min: 0,
+              max: 1,
+              step: 0.1,
+              description: t('Select border width'),
+            },
```
```suggestion
          {
            name: 'border_width',
            config: {
              type: 'SliderControl',
              label: t('Border width'),
              renderTrigger: true,
              min: 0,
              max: 2,
              default: 0,
              step: 0.1,
              description: t('The width of the elements border'),
            },
```

michael-s-molina commented on superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/transformProps.ts:
```diff
@@ -86,6 +89,8 @@ export default function transformProps(
     metric = '',
     normalizeAcross,
     normalized,
+    elementBorder = { r: 255, g: 255, b: 255, a: 1 },
+    bordersWidth = 0,
```
```suggestion
    borderColor,
    borderWidth,
```

michael-s-molina commented on superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/transformProps.ts:
```diff
@@ -161,6 +166,20 @@ export default function transformProps(
           return valueFormatter(paramsValue?.[2] as number | null | undefined);
         },
       },
+      itemStyle: {
+        borderColor: addAlpha(
+          rgbToHex(elementBorder.r, elementBorder.g, elementBorder.b),
+          elementBorder.a,
+        ),
+        borderWidth: bordersWidth,
+      },
```
```suggestion
      itemStyle: {
        borderColor: addAlpha(
          rgbToHex(borderColor.r, borderColor.g, borderColor.b),
          borderColor.a,
        ),
        borderWidth,
      },
```

michael-s-molina commented on superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/types.ts:
```diff
@@ -33,6 +34,8 @@ export interface HeatmapFormData extends QueryFormData {
   metric: QueryFormMetric;
   normalizeAcross: 'heatmap' | 'x' | 'y';
   normalized?: boolean;
+  elementBorder?: RgbaColor;
+  bordersWidth: number;
```
```suggestion
  borderColor: RgbaColor;
  borderWidth: number;
```

michael-s-molina reviewed (commented):
Thanks for the changes @Dev10-34. I left some suggestion to improve variable names.

Dev10-34 committed:
Update superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/controlPanel.tsx

Co-authored-by: Michael S. Molina <70410625+michael-s-molina@users.noreply.github.com>

Dev10-34 committed:
Update superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/controlPanel.tsx

Co-authored-by: Michael S. Molina <70410625+michael-s-molina@users.noreply.github.com>

Dev10-34 committed:
Update superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/controlPanel.tsx

Co-authored-by: Michael S. Molina <70410625+michael-s-molina@users.noreply.github.com>

Dev10-34 committed:
Update superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/controlPanel.tsx

Co-authored-by: Michael S. Molina <70410625+michael-s-molina@users.noreply.github.com>

Dev10-34 committed:
Update superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/transformProps.ts

Co-authored-by: Michael S. Molina <70410625+michael-s-molina@users.noreply.github.com>

Dev10-34 committed:
Update superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/transformProps.ts

Co-authored-by: Michael S. Molina <70410625+michael-s-molina@users.noreply.github.com>

Dev10-34 committed:
Update superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/types.ts

Co-authored-by: Michael S. Molina <70410625+michael-s-molina@users.noreply.github.com>

Dev10-34 commented:
I applied all the suggestions which you gave, however, upon re running there seems to be a bug sorta where even though the border width is set to 0 as default, on initial creation of the chart. If you move the border width slider, it fixes itself. Everything else works as intended functionally, just the visuals of the initial chart creation shows the borders as well, even if all the defaults are set correctly.
Giving borderWidth variable in transformProps.ts an initial value as 0 seems to fix this issue

Dev10-34 committed:
modified:   superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/transformProps.ts

Dev10-34 commented on superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/transformProps.ts:
```diff
@@ -86,6 +89,8 @@ export default function transformProps(
     normalizeAcross,
     normalized,
+    borderColor,
+    borderWidth = 0,
```
This fixes the issue

michael-s-molina commented on superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/controlPanel.tsx:
```diff
@@ -224,6 +224,31 @@ const config: ControlPanelConfig = {
         ['y_axis_format'],
         ['x_axis_time_format'],
         ['currency_format'],
+        [
```
Can you move these configs after `linear_color_scheme`? It makes sense to keep close the color schemes.

Dev10-34 replied:
I have made the suggested changes

Dev10-34 committed:
modified:   superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/controlPanel.tsx

michael-s-molina reviewed (approved):
Thank you so much for this great improvement @Dev10-34 and congrats on your first Superset PR!

Dev10-34 commented:
> Thank you so much for this great improvement @Dev10-34 and congrats on your first Superset PR!

Thank you so much for having me!! Is there any possibility that these changes will be reflected in future tags/releases?

michael-s-molina commented:
> Is there any possibility that these changes will be reflected in future tags/releases?

It will be present in the latest Docker tag now and it will be officially released in 5.1.

Dev10-34 commented:
> > Is there any possibility that these changes will be reflected in future tags/releases?
> 
> The `latest` Docker tag will contain this feature. It will be officially released in 5.1.

Glad to have been of help! Looking forward to more commits!
